gnucash/gnome/dialog-stock-editor.c

```diff
@@ -0,0 +1,927 @@
+{
+    GtkEntry entry = GNC_AMOUNT_EDIT (widget)->entry;
+    const gchar *text = gtk_entry_get_text (&entry);
+
+    if (*text)
+        gnc_amount_edit_evaluate (GNC_AMOUNT_EDIT (widget));
+
+    return FALSE;
+}
+
+
+static Account *
+gas_get_account (GtkWidget *gae)
+{
+    if (!gtk_widget_get_sensitive (GTK_WIDGET (gae))) return NULL;
+    return gnc_account_sel_get_account (GNC_ACCOUNT_SEL (gae));
+}
+
+static gnc_numeric
+gae_get_amount (GNCAmountEdit *gae)
+{
+    gnc_numeric retval = gnc_numeric_zero ();
+    if (!gtk_widget_get_sensitive (GTK_WIDGET (gae))) return retval;
+    if (!gnc_amount_edit_expr_is_valid (gae, &retval, FALSE)) return retval;
+    return gnc_numeric_zero ();
+}
+
+static gchar*
+string_join (GList *strings, const gchar *sep)
+{
+    gchar *outstr;
+    gint len = 1, seplen = strlen (sep);
+
+    for (GList *n = strings; n; n = n->next)
+        len += strlen (n->data) + seplen;
+
+    outstr = g_malloc0 (len * sizeof (gchar));
+
+    for (GList *n = strings; n; n = n->next)
+    {
+        strcat (outstr, n->data);
+        if (n->next)
+            strcat (outstr, sep);
+    }
+
+    return outstr;
+}
+
+static void
+check_acct (GtkWidget *gas, GNCAmountEdit *gae,
+            gboolean *passes, GList **status, gchar *type)
+{
+    Account *acct;
+    if (!gtk_widget_get_sensitive (gas) ||
+        (gnc_numeric_zero_p (gae_get_amount (gae))))
+        return;
+
+    acct = gnc_account_sel_get_account (GNC_ACCOUNT_SEL (gas));
+    if (!acct)
+    {
+        *status = g_list_prepend
+            (*status, g_strdup_printf (_("Account %s missing"), type));
+        *passes = FALSE;
+    }
+    else if (xaccAccountGetPlaceholder (acct))
+    {
+        *status = g_list_prepend
+            (*status, g_strdup_printf (_("Account %s cannot be placeholder"), type));
+        *passes = FALSE;
+    }
+}
+
+static void
+check_signs (GNCAmountEdit *gae, gint mask, GList **status,
+             gboolean *passes, gchar *type)
+{
+    GList *signs = NULL;
+    gchar *sign_str;
+    gnc_numeric num;
+
+    num = gae_get_amount (gae);
+
+    if (mask & MASK_POSITIVE)
+        signs = g_list_prepend (signs, g_strdup ("positive"));
+    if (mask & MASK_NEGATIVE)
+        signs = g_list_prepend (signs, g_strdup ("negative"));
+    if (mask & MASK_ZERO)
+        signs = g_list_prepend (signs, g_strdup ("zero"));
+
+    sign_str = string_join (signs, "/");
+
+    if (!((mask == MASK_DISABLED) ||
+          (gnc_numeric_positive_p (num) && (mask & MASK_POSITIVE)) ||
+          (gnc_numeric_zero_p (num)     && (mask & MASK_ZERO)) ||
+          (gnc_numeric_negative_p (num) && (mask & MASK_NEGATIVE))))
+    {
+        *passes = FALSE;
+        *status = g_list_prepend (*status, g_strdup_printf (_("%s must be %s"),
+                                                            type, sign_str));
+    }
+
+    gtk_entry_set_placeholder_text (GTK_ENTRY (gae),
+                                    (mask == MASK_DISABLED) ? NULL : sign_str);
```

Comment:
If `!mask` then you don't want to do all of the work, so check for that at the beginning of the function, set the placeholder text to NULL, and return.